// cosmos-client/types/address/address.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Prefix;
(function (Prefix) {
    Prefix["Cosmos"] = "cosmos";
    Prefix["Public"] = "pub";
    Prefix["Account"] = "acc";
    Prefix["Validator"] = "val";
    Prefix["Operator"] = "oper";
    Prefix["Consensus"] = "cons";
})(Prefix = exports.Prefix || (exports.Prefix = {}));
exports.bech32Prefix = {
    accAddr: Prefix.Cosmos,
    accPub: Prefix.Cosmos + Prefix.Public,
    valAddr: Prefix.Cosmos + Prefix.Validator + Prefix.Operator,
    valPub: Prefix.Cosmos + Prefix.Validator + Prefix.Operator + Prefix.Public,
    consAddr: Prefix.Cosmos + Prefix.Validator + Prefix.Consensus,
    consPub: Prefix.Cosmos + Prefix.Validator + Prefix.Consensus + Prefix.Public,
};
/**
 * 各種アドレスの基底クラス。
 */
var Address = /** @class */ (function () {
    /**
     *
     * @param value
     */
    function Address(value) {
        var addressLength = 20;
        if (value.length !== addressLength) {
            throw Error("Address must be 20 bytes length.");
        }
        this._value = value;
    }
    /**
     *
     * @param pubKey
     */
    Address.fromPublicKey = function (pubKey) {
        return new Address(pubKey.getAddress());
    };
    /**
     *
     * @param accAddr
     * @param accPub
     * @param valAddr
     * @param valPub
     * @param consAddr
     * @param consPub
     */
    Address.setBech32Prefix = function (accAddr, accPub, valAddr, valPub, consAddr, consPub) {
        exports.bech32Prefix.accAddr = accAddr;
        exports.bech32Prefix.accPub = accPub;
        exports.bech32Prefix.valAddr = valAddr;
        exports.bech32Prefix.valPub = valPub;
        exports.bech32Prefix.consAddr = consAddr;
        exports.bech32Prefix.consPub = consPub;
    };
    return Address;
}());
exports.Address = Address;
