// cosmos-client/x/bank/bank.spec.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var __1 = require("../../");
var auth_1 = require("../auth");
var bank_1 = require("../bank");
var tendermint_1 = require("../../tendermint");
test("bank", function () { return __awaiter(void 0, void 0, void 0, function () {
    var sdk, privKeyBuffer, privKey, fromAddress, account, toAddress, unsignedStdTx, signedStdTx, result;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0:
                sdk = new __1.CosmosSDK("", "test");
                privKeyBuffer = Buffer.from("36d1043c6e23eb15c928da41043bfd183b6ce13f9e592c9a45ac431c4a08b924", "hex");
                privKey = new tendermint_1.PrivKeySecp256k1(privKeyBuffer);
                fromAddress = __1.AccAddress.fromPublicKey(privKey.getPubKey());
                return [4 /*yield*/, auth_1.auth
                        .accountsAddressGet(sdk, fromAddress)
                        .then(function (res) { return res.data.result; })];
            case 1:
                account = _a.sent();
                toAddress = fromAddress;
                return [4 /*yield*/, bank_1.bank
                        .accountsAddressTransfersPost(sdk, toAddress, {
                        base_req: {
                            from: fromAddress.toBech32(),
                            memo: "Hello, world!",
                            chain_id: sdk.chainID,
                            account_number: account.account_number.toString(),
                            sequence: account.sequence.toString(),
                            gas: "",
                            gas_adjustment: "",
                            fees: [],
                            simulate: false,
                        },
                        amount: [{ denom: "token", amount: "1000" }],
                    })
                        .then(function (res) { return res.data; })];
            case 2:
                unsignedStdTx = _a.sent();
                signedStdTx = auth_1.auth.signStdTx(sdk, privKey, unsignedStdTx, account.account_number.toString(), account.sequence.toString());
                return [4 /*yield*/, auth_1.auth
                        .txsPost(sdk, signedStdTx, "sync")
                        .then(function (res) { return res.data; })];
            case 3:
                result = _a.sent();
                console.log(result);
                return [2 /*return*/];
        }
    });
}); });
