// @xchainjs/xchain-binance/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var client = require('@binance-chain/javascript-sdk/lib/client');
var crypto = require('@binance-chain/javascript-sdk/lib/crypto');
var xchainClient = require('@xchainjs/xchain-client');
var xchainUtil = require('@xchainjs/xchain-util');
var axios = require('axios');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var axios__default = /*#__PURE__*/_interopDefaultLegacy(axios);

/*! *****************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */
/* global Reflect, Promise */

var extendStatics = function(d, b) {
    extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
    return extendStatics(d, b);
};

function __extends(d, b) {
    if (typeof b !== "function" && b !== null)
        throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
    extendStatics(d, b);
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
}

function __awaiter(thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
}

function __generator(thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
}

/**
 * Type guard for runtime checks of `Fee`
 *
 * @param {Fee|TransferFee|DexFees} v
 * @returns {boolean} `true` or `false`.
 */
var isFee = function (v) { var _a, _b, _c; return !!((_a = v) === null || _a === void 0 ? void 0 : _a.msg_type) && ((_b = v) === null || _b === void 0 ? void 0 : _b.fee) !== undefined && ((_c = v) === null || _c === void 0 ? void 0 : _c.fee_for) !== undefined; };
/**
 * Type guard for `TransferFee`
 *
 * @param {Fee|TransferFee|DexFees} v
 * @returns {boolean} `true` or `false`.
 */
var isTransferFee = function (v) { var _a, _b; return isFee((_a = v) === null || _a === void 0 ? void 0 : _a.fixed_fee_params) && !!((_b = v) === null || _b === void 0 ? void 0 : _b.multi_transfer_fee); };
/**
 * Type guard for `Account`
 *
 * @param {unknown} v
 * @returns {boolean} `true` or `false`.
 */
var isAccount = function (v) {
    return typeof v.account_number === 'number' &&
        typeof v.address === 'string' &&
        Array.isArray(v.balances) &&
        Array.isArray(v.public_key) &&
        typeof v.flags === 'number' &&
        typeof v.sequence === 'number';
};
/**
 * Get TxType
 *
 * @param {BinanceTxType} t
 * @returns {TxType} `transfer` or `unknown`.
 */
var getTxType = function (t) {
    if (t === 'TRANSFER' || t === 'DEPOSIT')
        return xchainClient.TxType.Transfer;
    return xchainClient.TxType.Unknown;
};
/**
 * Parse Tx
 *
 * @param {BinanceTx} t The transaction to be parsed. (optional)
 * @returns {Tx|null} The transaction parsed from the binance tx.
 */
var parseTx = function (tx) {
    var asset = xchainUtil.assetFromString(xchainUtil.AssetBNB.chain + "." + tx.txAsset);
    if (!asset)
        return null;
    return {
        asset: asset,
        from: [
            {
                from: tx.fromAddr,
                amount: xchainUtil.assetToBase(xchainUtil.assetAmount(tx.value, 8)),
            },
        ],
        to: [
            {
                to: tx.toAddr,
                amount: xchainUtil.assetToBase(xchainUtil.assetAmount(tx.value, 8)),
            },
        ],
        date: new Date(tx.timeStamp),
        type: getTxType(tx.txType),
        hash: tx.txHash,
    };
};
/**
 * Get DerivePath
 *
 * @param {number} index (optional)
 * @returns {DerivePath} The binance derivation path by the index.
 */
var getDerivePath = function (index) {
    if (index === void 0) { index = 0; }
    return [44, 714, 0, 0, index];
};
/**
 * Get the default fee.
 *
 * @returns {Fees} The default fee.
 */
var getDefaultFees = function () {
    return xchainClient.singleFee(xchainClient.FeeType.FlatFee, xchainUtil.baseAmount(37500));
};
/**
 * Get address prefix based on the network.
 *
 * @param {Network} network
 * @returns {string} The address prefix based on the network.
 *
 **/
var getPrefix = function (network) {
    switch (network) {
        case xchainClient.Network.Mainnet:
            return 'bnb';
        case xchainClient.Network.Testnet:
            return 'tbnb';
    }
};

/**
 * Custom Binance client
 */
var Client = /** @class */ (function (_super) {
    __extends(Client, _super);
    /**
     * Constructor
     *
     * Client has to be initialised with network type and phrase.
     * It will throw an error if an invalid phrase has been passed.
     *
     * @param {XChainClientParams} params
     *
     * @throws {"Invalid phrase"} Thrown if the given phase is invalid.
     */
    function Client(params) {
        var _this = _super.call(this, xchainUtil.Chain.Binance, params) || this;
        _this.bncClient = new client.BncClient(_this.getClientUrl());
        _this.bncClient.chooseNetwork(_this.getNetwork());
        return _this;
    }
    /**
     * Get the BncClient interface.
     *
     * @returns {BncClient} The BncClient from `@binance-chain/javascript-sdk`.
     */
    Client.prototype.getBncClient = function () {
        return this.bncClient;
    };
    /**
     * Set/update the current network.
     *
     * @param {Network} network
     * @returns {void}
     *
     * @throws {"Network must be provided"}
     * Thrown if network has not been set before.
     */
    Client.prototype.setNetwork = function (network) {
        _super.prototype.setNetwork.call(this, network);
        this.bncClient = new client.BncClient(this.getClientUrl());
        this.bncClient.chooseNetwork(network);
    };
    /**
     * Get the client url.
     *
     * @returns {string} The client url for binance chain based on the network.
     */
    Client.prototype.getClientUrl = function () {
        switch (this.network) {
            case xchainClient.Network.Mainnet:
                return 'https://dex.binance.org';
            case xchainClient.Network.Testnet:
                return 'https://testnet-dex.binance.org';
        }
    };
    /**
     * Get the explorer url.
     *
     * @returns {string} The explorer url based on the network.
     */
    Client.prototype.getExplorerUrl = function () {
        switch (this.network) {
            case xchainClient.Network.Mainnet:
                return 'https://explorer.binance.org';
            case xchainClient.Network.Testnet:
                return 'https://testnet-explorer.binance.org';
        }
    };
    /**
     * Get the explorer url for the given address.
     *
     * @param {Address} address
     * @returns {string} The explorer url for the given address based on the network.
     */
    Client.prototype.getExplorerAddressUrl = function (address) {
        return this.getExplorerUrl() + "/address/" + address;
    };
    /**
     * Get the explorer url for the given transaction id.
     *
     * @param {string} txID
     * @returns {string} The explorer url for the given transaction id based on the network.
     */
    Client.prototype.getExplorerTxUrl = function (txID) {
        return this.getExplorerUrl() + "/tx/" + txID;
    };
    /**
     * @private
     * Get private key.
     *
     * @param {number} index account index for the derivation path
     * @returns {PrivKey} The privkey generated from the given phrase
     *
     * @throws {"Phrase not set"}
     * Throws an error if phrase has not been set before
     * */
    Client.prototype.getPrivateKey = function (index) {
        if (!this.phrase)
            throw new Error('Phrase not set');
        return crypto.getPrivateKeyFromMnemonic(this.phrase, true, index);
    };
    /**
     * Get the current address.
     *
     * @param {number} index (optional) Account index for the derivation path
     * @returns {Address} The current address.
     *
     * @throws {Error} Thrown if phrase has not been set before. A phrase is needed to create a wallet and to derive an address from it.
     */
    Client.prototype.getAddress = function (index) {
        if (index === void 0) { index = 0; }
        return crypto.getAddressFromPrivateKey(this.getPrivateKey(index), getPrefix(this.network));
    };
    /**
     * Validate the given address.
     *
     * @param {Address} address
     * @returns {boolean} `true` or `false`
     */
    Client.prototype.validateAddress = function (address) {
        return this.bncClient.checkAddress(address, getPrefix(this.network));
    };
    /**
     * Get account data of wallets or by given address.
     *
     * @param {Address} address (optional) By default, it will return account data of current wallet.
     * @param {number} index (optional) Account index for the derivation path
     *
     * @returns {Account} account details of given address.
     */
    Client.prototype.getAccount = function (address, index) {
        if (index === void 0) { index = 0; }
        return __awaiter(this, void 0, void 0, function () {
            var accountAddress, response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        accountAddress = address || this.getAddress(index);
                        return [4 /*yield*/, this.bncClient.getAccount(accountAddress)];
                    case 1:
                        response = _a.sent();
                        if (!response || !response.result || !isAccount(response.result))
                            return [2 /*return*/, Promise.reject(Error("Could not get account data for address " + accountAddress))];
                        return [2 /*return*/, response.result];
                }
            });
        });
    };
    /**
     * Get the balance of a given address.
     *
     * @param {Address} address By default, it will return the balance of the current wallet. (optional)
     * @param {Asset} asset If not set, it will return all assets available. (optional)
     * @returns {Balance[]} The balance of the address.
     */
    Client.prototype.getBalance = function (address, assets) {
        return __awaiter(this, void 0, void 0, function () {
            var balances;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.bncClient.getBalance(address)];
                    case 1:
                        balances = _a.sent();
                        return [2 /*return*/, balances
                                .map(function (balance) {
                                return {
                                    asset: xchainUtil.assetFromString(xchainUtil.Chain.Binance + "." + balance.symbol) || xchainUtil.AssetBNB,
                                    amount: xchainUtil.assetToBase(xchainUtil.assetAmount(balance.free, 8)),
                                };
                            })
                                .filter(function (balance) { return !assets || assets.filter(function (asset) { return xchainUtil.assetToString(balance.asset) === xchainUtil.assetToString(asset); }).length; })];
                }
            });
        });
    };
    /**
     * @private
     * Search transactions with parameters.
     *
     * @returns {Params} The parameters to be used for transaction search.
     * */
    Client.prototype.searchTransactions = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var clientUrl, url, endTime, diffTime, key, value, txHistory;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        clientUrl = this.getClientUrl() + "/api/v1/transactions";
                        url = new URL(clientUrl);
                        endTime = Date.now();
                        diffTime = 90 * 24 * 60 * 60 * 1000;
                        url.searchParams.set('endTime', endTime.toString());
                        url.searchParams.set('startTime', (endTime - diffTime).toString());
                        for (key in params) {
                            value = params[key];
                            if (value) {
                                url.searchParams.set(key, value);
                                if (key === 'startTime' && !params['endTime']) {
                                    url.searchParams.set('endTime', (parseInt(value) + diffTime).toString());
                                }
                                if (key === 'endTime' && !params['startTime']) {
                                    url.searchParams.set('startTime', (parseInt(value) - diffTime).toString());
                                }
                            }
                        }
                        return [4 /*yield*/, axios__default['default'].get(url.toString())];
                    case 1:
                        txHistory = (_a.sent()).data;
                        return [2 /*return*/, {
                                total: txHistory.total,
                                txs: txHistory.tx.map(parseTx).filter(Boolean),
                            }];
                }
            });
        });
    };
    /**
     * Get transaction history of a given address with pagination options.
     * By default it will return the transaction history of the current wallet.
     *
     * @param {TxHistoryParams} params The options to get transaction history. (optional)
     * @returns {TxsPage} The transaction history.
     */
    Client.prototype.getTransactions = function (params) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0: return [4 /*yield*/, this.searchTransactions({
                            address: params && params.address,
                            limit: params && ((_a = params.limit) === null || _a === void 0 ? void 0 : _a.toString()),
                            offset: params && ((_b = params.offset) === null || _b === void 0 ? void 0 : _b.toString()),
                            startTime: params && params.startTime && params.startTime.getTime().toString(),
                            txAsset: params && params.asset,
                        })];
                    case 1: return [2 /*return*/, _c.sent()];
                }
            });
        });
    };
    /**
     * Get the transaction details of a given transaction id.
     *
     * @param {string} txId The transaction id.
     * @returns {Tx} The transaction details of the given transaction id.
     */
    Client.prototype.getTransactionData = function (txId) {
        return __awaiter(this, void 0, void 0, function () {
            var txResult, blockHeight, address, msgs, msg, txHistory, transaction;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, axios__default['default'].get(this.getClientUrl() + "/api/v1/tx/" + txId + "?format=json")];
                    case 1:
                        txResult = (_a.sent()).data;
                        blockHeight = txResult.height;
                        address = '';
                        msgs = txResult.tx.value.msg;
                        if (msgs.length) {
                            msg = msgs[0].value;
                            if (msg.inputs && msg.inputs.length) {
                                address = msg.inputs[0].address;
                            }
                            else if (msg.outputs && msg.outputs.length) {
                                address = msg.outputs[0].address;
                            }
                        }
                        return [4 /*yield*/, this.searchTransactions({ address: address, blockHeight: blockHeight })];
                    case 2:
                        txHistory = _a.sent();
                        transaction = txHistory.txs.filter(function (tx) { return tx.hash === txId; })[0];
                        if (!transaction) {
                            throw new Error('transaction not found');
                        }
                        return [2 /*return*/, transaction];
                }
            });
        });
    };
    /**
     * Broadcast multi-send transaction.
     *
     * @param {MultiSendParams} params The multi-send transfer options.
     * @returns {TxHash} The transaction hash.
     */
    Client.prototype.multiSend = function (_a) {
        var _b = _a.walletIndex, walletIndex = _b === void 0 ? 0 : _b, transactions = _a.transactions, _c = _a.memo, memo = _c === void 0 ? '' : _c;
        return __awaiter(this, void 0, void 0, function () {
            var derivedAddress, transferResult;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        derivedAddress = this.getAddress(walletIndex);
                        return [4 /*yield*/, this.bncClient.initChain()];
                    case 1:
                        _d.sent();
                        return [4 /*yield*/, this.bncClient.setPrivateKey(this.getPrivateKey(walletIndex))];
                    case 2:
                        _d.sent();
                        return [4 /*yield*/, this.bncClient.multiSend(derivedAddress, transactions.map(function (transaction) {
                                return {
                                    to: transaction.to,
                                    coins: transaction.coins.map(function (coin) {
                                        return {
                                            denom: coin.asset.symbol,
                                            amount: xchainUtil.baseToAsset(coin.amount).amount().toString(),
                                        };
                                    }),
                                };
                            }), memo)];
                    case 3:
                        transferResult = _d.sent();
                        return [2 /*return*/, transferResult.result.map(function (txResult) { var _a; return (_a = txResult === null || txResult === void 0 ? void 0 : txResult.hash) !== null && _a !== void 0 ? _a : ''; })[0]];
                }
            });
        });
    };
    /**
     * Transfer balances.
     *
     * @param {TxParams} params The transfer options.
     * @returns {TxHash} The transaction hash.
     */
    Client.prototype.transfer = function (_a) {
        var walletIndex = _a.walletIndex, asset = _a.asset, amount = _a.amount, recipient = _a.recipient, memo = _a.memo;
        return __awaiter(this, void 0, void 0, function () {
            var transferResult;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.bncClient.initChain()];
                    case 1:
                        _b.sent();
                        return [4 /*yield*/, this.bncClient.setPrivateKey(this.getPrivateKey(walletIndex || 0))];
                    case 2:
                        _b.sent();
                        return [4 /*yield*/, this.bncClient.transfer(this.getAddress(walletIndex), recipient, xchainUtil.baseToAsset(amount).amount().toString(), asset ? asset.symbol : xchainUtil.AssetBNB.symbol, memo)];
                    case 3:
                        transferResult = _b.sent();
                        return [2 /*return*/, transferResult.result.map(function (txResult) { var _a; return (_a = txResult === null || txResult === void 0 ? void 0 : txResult.hash) !== null && _a !== void 0 ? _a : ''; })[0]];
                }
            });
        });
    };
    /**
     * Get the current transfer fee.
     *
     * @returns {TransferFee} The current transfer fee.
     */
    Client.prototype.getTransferFee = function () {
        return __awaiter(this, void 0, void 0, function () {
            var feesArray, transferFee;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, axios__default['default'].get(this.getClientUrl() + "/api/v1/fees")];
                    case 1:
                        feesArray = (_a.sent()).data;
                        transferFee = feesArray.filter(isTransferFee)[0];
                        if (!transferFee)
                            throw new Error('failed to get transfer fees');
                        return [2 /*return*/, transferFee];
                }
            });
        });
    };
    /**
     * Get the current fee.
     *
     * @returns {Fees} The current fee.
     */
    Client.prototype.getFees = function () {
        return __awaiter(this, void 0, void 0, function () {
            var singleTxFee, _a, error_1, transferFee;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        singleTxFee = undefined;
                        _b.label = 1;
                    case 1:
                        _b.trys.push([1, 3, , 4]);
                        _a = xchainUtil.baseAmount;
                        return [4 /*yield*/, this.getFeeRateFromThorchain()];
                    case 2:
                        singleTxFee = _a.apply(void 0, [_b.sent()]);
                        return [3 /*break*/, 4];
                    case 3:
                        error_1 = _b.sent();
                        console.log(error_1);
                        console.warn("Error pulling rates from thorchain, will try alternate");
                        return [3 /*break*/, 4];
                    case 4:
                        if (!!singleTxFee) return [3 /*break*/, 6];
                        return [4 /*yield*/, this.getTransferFee()];
                    case 5:
                        transferFee = _b.sent();
                        singleTxFee = xchainUtil.baseAmount(transferFee.fixed_fee_params.fee);
                        _b.label = 6;
                    case 6: return [2 /*return*/, xchainClient.singleFee(xchainClient.FeeType.FlatFee, singleTxFee)];
                }
            });
        });
    };
    /**
     * Get the current fee for multi-send transaction.
     *
     * @returns {Fees} The current fee for multi-send transaction.
     */
    Client.prototype.getMultiSendFees = function () {
        return __awaiter(this, void 0, void 0, function () {
            var transferFee, multiTxFee;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getTransferFee()];
                    case 1:
                        transferFee = _a.sent();
                        multiTxFee = xchainUtil.baseAmount(transferFee.multi_transfer_fee);
                        return [2 /*return*/, {
                                type: 'base',
                                average: multiTxFee,
                                fast: multiTxFee,
                                fastest: multiTxFee,
                            }];
                }
            });
        });
    };
    /**
     * Get the current fee for both single and multi-send transaction.
     *
     * @returns {SingleAndMultiFees} The current fee for both single and multi-send transaction.
     */
    Client.prototype.getSingleAndMultiFees = function () {
        return __awaiter(this, void 0, void 0, function () {
            var transferFee, singleTxFee, multiTxFee;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getTransferFee()];
                    case 1:
                        transferFee = _a.sent();
                        singleTxFee = xchainUtil.baseAmount(transferFee.fixed_fee_params.fee);
                        multiTxFee = xchainUtil.baseAmount(transferFee.multi_transfer_fee);
                        return [2 /*return*/, {
                                single: {
                                    type: 'base',
                                    fast: singleTxFee,
                                    fastest: singleTxFee,
                                    average: singleTxFee,
                                },
                                multi: {
                                    type: 'base',
                                    average: multiTxFee,
                                    fast: multiTxFee,
                                    fastest: multiTxFee,
                                },
                            }];
                }
            });
        });
    };
    return Client;
}(xchainClient.BaseXChainClient));

/**
 * Order status as part of an order
 * See description of Order.status for more detail https://docs.binance.org/api-reference/dex-api/paths.html#order
 */
(function (OrderStatus) {
    OrderStatus["Ack"] = "Ack";
    OrderStatus["PartialFill"] = "PartialFill";
    OrderStatus["IocNoFill"] = "IocNoFill";
    OrderStatus["FullyFill"] = "FullyFill";
    OrderStatus["Canceled"] = "Canceled";
    OrderStatus["Expired"] = "Expired";
    OrderStatus["FailedBlocking"] = "FailedBlocking";
    OrderStatus["FailedMatching"] = "FailedMatching";
    OrderStatus["IocExpire"] = "IocExpire";
})(exports.OrderStatus || (exports.OrderStatus = {}));

/**
 * Type definitions for data of Binance WebSocket Streams
 * @see https://docs.binance.org/api-reference/dex-api/ws-streams.html
 *
 */
/**
 * Taker (as part of {@link Trade})
 */
var Taker;
(function (Taker) {
    Taker[Taker["UNKNOWN"] = 0] = "UNKNOWN";
    Taker[Taker["SELL_TAKER"] = 1] = "SELL_TAKER";
    Taker[Taker["BUY_TAKER"] = 2] = "BUY_TAKER";
    Taker[Taker["BUY_SURPLUS"] = 3] = "BUY_SURPLUS";
    Taker[Taker["SELL_SURPLUS"] = 4] = "SELL_SURPLUS";
    Taker[Taker["NEUTRAL"] = 5] = "NEUTRAL";
})(Taker || (Taker = {}));

var binanceWs = /*#__PURE__*/Object.freeze({
    __proto__: null,
    get Taker () { return Taker; }
});

exports.Client = Client;
exports.WS = binanceWs;
exports.getDefaultFees = getDefaultFees;
exports.getDerivePath = getDerivePath;
exports.getPrefix = getPrefix;
