// cosmos-client/x/staking/module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var api_1 = require("../../api");
var codec_1 = require("../../codec");
function delegatorsDelegatorAddrDelegationsGet(sdk, address) {
    return sdk.wrapResponseWithHeight(new api_1.StakingApi(undefined, sdk.url).stakingDelegatorsDelegatorAddrDelegationsGet(address.toBech32()));
}
exports.delegatorsDelegatorAddrDelegationsGet = delegatorsDelegatorAddrDelegationsGet;
function delegatorsDelegatorAddrDelegationsPost(sdk, delegator) {
    return new api_1.StakingApi(undefined, sdk.url)
        .stakingDelegatorsDelegatorAddrDelegationsPost(delegator.toBech32())
        .then(function (res) {
        res.data = codec_1.codec.fromJSONString(JSON.stringify(res.data));
        return res;
    });
}
exports.delegatorsDelegatorAddrDelegationsPost = delegatorsDelegatorAddrDelegationsPost;
function delegatorsDelegatorAddrDelegationsValidatorAddrGet(sdk, delegator, validator) {
    return sdk.wrapResponseWithHeight(new api_1.StakingApi(undefined, sdk.url).stakingDelegatorsDelegatorAddrDelegationsValidatorAddrGet(delegator.toBech32(), validator.toBech32()));
}
exports.delegatorsDelegatorAddrDelegationsValidatorAddrGet = delegatorsDelegatorAddrDelegationsValidatorAddrGet;
function delegatorsDelegatorAddrRedelegationsPost(sdk, delegator) {
    return new api_1.StakingApi(undefined, sdk.url)
        .stakingDelegatorsDelegatorAddrRedelegationsPost(delegator.toBech32())
        .then(function (res) {
        res.data = codec_1.codec.fromJSONString(JSON.stringify(res.data));
        return res;
    });
}
exports.delegatorsDelegatorAddrRedelegationsPost = delegatorsDelegatorAddrRedelegationsPost;
function delegatorsDelegatorAddrUnbondingDelegationsGet(sdk, delegator) {
    return sdk.wrapResponseWithHeight(new api_1.StakingApi(undefined, sdk.url).stakingDelegatorsDelegatorAddrUnbondingDelegationsGet(delegator.toBech32()));
}
exports.delegatorsDelegatorAddrUnbondingDelegationsGet = delegatorsDelegatorAddrUnbondingDelegationsGet;
function delegatorsDelegatorAddrUnbondingDelegationsPost(sdk, delegator) {
    return new api_1.StakingApi(undefined, sdk.url)
        .stakingDelegatorsDelegatorAddrUnbondingDelegationsPost(delegator.toBech32())
        .then(function (res) {
        res.data = codec_1.codec.fromJSONString(JSON.stringify(res.data));
        return res;
    });
}
exports.delegatorsDelegatorAddrUnbondingDelegationsPost = delegatorsDelegatorAddrUnbondingDelegationsPost;
function delegatorsDelegatorAddrUnbondingDelegationsValidatorAddrGet(sdk, delegator, validator) {
    return sdk.wrapResponseWithHeight(new api_1.StakingApi(undefined, sdk.url).stakingDelegatorsDelegatorAddrUnbondingDelegationsValidatorAddrGet(delegator.toBech32(), validator.toBech32()));
}
exports.delegatorsDelegatorAddrUnbondingDelegationsValidatorAddrGet = delegatorsDelegatorAddrUnbondingDelegationsValidatorAddrGet;
function delegatorsDelegatorAddrValidatorsGet(sdk, delegator) {
    return sdk.wrapResponseWithHeight(new api_1.StakingApi(undefined, sdk.url).stakingDelegatorsDelegatorAddrValidatorsGet(delegator.toBech32()));
}
exports.delegatorsDelegatorAddrValidatorsGet = delegatorsDelegatorAddrValidatorsGet;
function delegatorsDelegatorAddrValidatorsValidatorAddrGet(sdk, delegator, validator) {
    return sdk.wrapResponseWithHeight(new api_1.StakingApi(undefined, sdk.url).stakingDelegatorsDelegatorAddrValidatorsValidatorAddrGet(delegator.toBech32(), validator.toBech32()));
}
exports.delegatorsDelegatorAddrValidatorsValidatorAddrGet = delegatorsDelegatorAddrValidatorsValidatorAddrGet;
function parametersGet(sdk) {
    return sdk.wrapResponseWithHeight(new api_1.StakingApi(undefined, sdk.url).stakingParametersGet());
}
exports.parametersGet = parametersGet;
function poolGet(sdk) {
    return sdk.wrapResponseWithHeight(new api_1.StakingApi(undefined, sdk.url).stakingPoolGet());
}
exports.poolGet = poolGet;
function redelegationsGet(sdk, delegator, validatorFrom, validatorTo) {
    return sdk.wrapResponseWithHeight(new api_1.StakingApi(undefined, sdk.url).stakingRedelegationsGet(delegator === null || delegator === void 0 ? void 0 : delegator.toBech32(), validatorFrom === null || validatorFrom === void 0 ? void 0 : validatorFrom.toBech32(), validatorTo === null || validatorTo === void 0 ? void 0 : validatorTo.toBech32()));
}
exports.redelegationsGet = redelegationsGet;
function validatorsGet(sdk, status, page, limit) {
    return sdk.wrapResponseWithHeight(new api_1.StakingApi(undefined, sdk.url).stakingValidatorsGet(status, page, limit));
}
exports.validatorsGet = validatorsGet;
function validatorsValidatorAddrDelegationsGet(sdk, validator) {
    return sdk.wrapResponseWithHeight(new api_1.StakingApi(undefined, sdk.url).stakingValidatorsValidatorAddrDelegationsGet(validator.toBech32()));
}
exports.validatorsValidatorAddrDelegationsGet = validatorsValidatorAddrDelegationsGet;
function validatorsValidatorAddrGet(sdk, validator) {
    return sdk.wrapResponseWithHeight(new api_1.StakingApi(undefined, sdk.url).stakingValidatorsValidatorAddrGet(validator.toBech32()));
}
exports.validatorsValidatorAddrGet = validatorsValidatorAddrGet;
function validatorsValidatorAddrUnbondingDelegationsGet(sdk, validator) {
    return sdk.wrapResponseWithHeight(new api_1.StakingApi(undefined, sdk.url).stakingValidatorsValidatorAddrUnbondingDelegationsGet(validator.toBech32()));
}
exports.validatorsValidatorAddrUnbondingDelegationsGet = validatorsValidatorAddrUnbondingDelegationsGet;
